From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mykyta Komarnytskyy <nkomarn@hotmail.com>
Date: Fri, 23 Oct 2020 09:20:01 -0700
Subject: [PATCH] Utilities

This patch includes all utilities required by the Hydrinity project and its patches.

This patch was created for the Hydrinity project <https://github.com/Hydrinity/Hydrinity> by Mykyta Komarnytskyy <nkomarn@hotmail.com> under the MIT license.

diff --git a/src/main/java/xyz/nkomarn/hydrinity/util/Constants.java b/src/main/java/xyz/nkomarn/hydrinity/util/Constants.java
new file mode 100644
index 0000000000000000000000000000000000000000..97ecbd1f0cf61ef565fa3554b75d8cc43ed6d201
--- /dev/null
+++ b/src/main/java/xyz/nkomarn/hydrinity/util/Constants.java
@@ -0,0 +1,7 @@
+package xyz.nkomarn.hydrinity.util;
+
+public class Constants {
+
+    public static final int[] EMPTY_ARRAY = new int[0];
+    public static final int[] ZERO_ARRAY = new int[]{0};
+}
diff --git a/src/main/java/xyz/nkomarn/hydrinity/util/MathUtils.java b/src/main/java/xyz/nkomarn/hydrinity/util/MathUtils.java
new file mode 100644
index 0000000000000000000000000000000000000000..fad859556c4091995339323991186159175dcd4e
--- /dev/null
+++ b/src/main/java/xyz/nkomarn/hydrinity/util/MathUtils.java
@@ -0,0 +1,17 @@
+package xyz.nkomarn.hydrinity.util;
+
+public class MathUtils {
+
+    /**
+     * An efficient modulo implementation that avoids the use of loops
+     * and instead uses faster bitwise operations to achieve the fastest
+     * possible execution time.
+     *
+     * @param dividend the dividend
+     * @param divisor the divisor
+     * @return the remainder
+     */
+    public static int fastMod(int dividend, int divisor) {
+        return dividend & (divisor - 1);
+    }
+}
